### ISSUE TITLE
Handling Binary Data

### ISSUE THREAD

**Issue by grouville** [dagger/dagger#5069]
## Background

While trying to implement examples for issue [Allow directories to be mounted as secrets](https://github.com/dagger/dagger/issues/4896), it became apparent that there are limitations when dealing with binary data in GraphQL using strings. This is relevant for the purpose of mounting directories as secrets, where binary data needs to be handled.

I encountered a problem when attempting to handle GPG files. The goal was to create and GPG-sign a built binary with Dagger by mounting the `~/.gnupg` directory into the container and running `gpg --detach-sign --armor binary-name`.

This required the following files:
- ~/.gnupg/pubring.kbx
- ~/.gnupg/trustdb.gpg
- ~/.gnupg/private-keys-v1.d/*

Handling binary data like GPG files in GraphQL led to errors due to improper escaping:
```shell
go run .                
Connected to engine c01f0fd17590
panic: input:1: Syntax Error GraphQL request (1:49) Invalid character escape sequence: \\x.
```

The initial idea was to implement a union type `SecretData = String | Bytes` in the `setSecret` query. However, according to the [discussion](https://github.com/dagger/dagger/issues/5000#issuecomment-1518352932), using unions as input values in GraphQL is not feasible.

## Issue

The main issue is finding a solution to handle both string and binary data when setting secrets in GraphQL, as the initial proposal of using a union type is not possible. This is important for mounting directories as secrets and handling binary data efficiently.

## Proposed Solution

Based on the comments and limitations mentioned in the discussion, the best solution seems to be to create a new function that accepts binary data as Base64 encoded string as input. This function would be a variation of withSecretVariable and handle binary data separately.

### Advantages
- No character encoding issues: By using Base64 encoding, we can ensure that all binary data is properly serialized and deserialized without any issues related to character encoding.
- Efficient data transfer: Base64-encoded binary data is more compact than a string representation, reducing the overhead associated with transferring binary data in GraphQL queries or mutations.
- Clear type information: Defining a custom type allows us to be explicit regarding the expected encoded format

#### Steps to implement the solution:

1. Create a new custom type `Base64` in the GraphQL schema to handle binary data.
```graphql
type Base64 string
```

2. Define a new function in the client SDK called WithSecretBytes that accepts an environment variable name and Base64 data as input.
```go
func (c *Container) setSecretBytes(name string, data Base64) *Container
```
3. Update the GraphQL schema to include the new function setSecretBytes.
```graphql
extend type Query {
  "Sets a secret given a user-defined name to its binary data and returns the secret."
  setSecretBytes(
    """
    The user-defined name for this secret
    """
    name: String!

    """
    The binary data of the secret
    """
    data: Base64!
  ): Secret!
}
```
4. Implement deserialization functions for the Base64 `setSecretBytes`, handling the decoding of binary from Base64.
5. Update the SDK example to show how to convert binary data to Base64-encoded strings before sending them as secret arguments.

By implementing this new function, we can effectively handle binary data in GraphQL, making it possible to mount directories as secrets in a more convenient and efficient manner.

cc @vikram-dagger @helderco @vito

**Comment by grouville** [comment]
We can also introduce a new parameter to the `setSecret` query that specifies the encoding format. This would allow you to handle different encoding formats more explicitly:
```graphql
enum EncodingFormat {
  STRING
  BASE64
}

extend type Query {
  """
  Sets a secret given a user-defined name to its data (either binary data or plaintext) and returns the secret.
  """
  setSecret(
    """
    The user-defined name for this secret
    """
    name: String!

    """
    The data of the secret (either binary data or plaintext)
    """
    data: String!

    """
    The encoding format of the secret data (defaults to STRING)
    """
    encoding: EncodingFormat = STRING
  ): Secret!
}
```
> (default do not seem to work though)

**Comment by vito** [comment]
> ```shell
> panic: input:1: Syntax Error GraphQL request (1:49) Invalid character escape sequence: \\x.
> ```

I think this is just because we took a shortcut in our query builder by assuming GraphQL strings follow the same rules as Go: https://github.com/dagger/dagger/pull/5008#discussion_r1178267836

We should probably update the query builders to follow the [GraphQL spec](https://github.com/graphql/graphql-spec/blob/main/spec/Section%202%20--%20Language.md#string-value) instead. I think with that change we shouldn't need to support alternative encodings, but I'm still in support of the approach in https://github.com/dagger/dagger/issues/5069#issuecomment-1531676095 if we want to go that route.

Worth noting GraphQL string escaping is something that would have to be implemented properly for every SDK. I'm not sure if this issue applies to the other SDKs.

**Comment by grouville** [comment]
> We should probably update the query builders to follow the [GraphQL spec](https://github.com/graphql/graphql-spec/blob/main/spec/Section%202%20--%20Language.md#string-value) instead.

Totally agree, thanks for insight 🙏 

> Worth noting GraphQL string escaping is something that would have to be implemented properly for every SDK. I'm not sure if this issue applies to the other SDKs.

Tested on Node SDK, it seems to work for one of the file on which the Go SDK breaks. I still have an issue on another file (bigger binary data), don't know if it is related or not. But I presume that the implementation is somewhat similar
```node.js
node --loader ts-node/esm test.mts
(node:25831) ExperimentalWarning: Custom ESM Loaders is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
file:///private/tmp/repro_mount_dir/node_modules/@dagger.io/dagger/dist/api/utils.js:5
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
                                                                ^
GraphQLRequestError: Error message
    at file:///private/tmp/repro_mount_dir/node_modules/@dagger.io/dagger/dist/api/utils.js:144:23
    at Generator.throw (<anonymous>)
    at rejected (file:///private/tmp/repro_mount_dir/node_modules/@dagger.io/dagger/dist/api/utils.js:5:65)
    at processTicksAndRejections (node:internal/process/task_queues:95:5) {
  cause: ClientError: invalid secret size 514282: {"response":{"data":null,"errors":[{"message":"invalid secret size 514282","locations":[{"line":2,"column":208}],"path":["container","from","withMountedSecret","withExec","stdout"]}],"status":200,"headers":{}},"request":{"query":"\n        { container  { from (address: \"alpine:3.17\") { withMountedSecret (path: \"/root/.gnupg/pubring.kbx\",source: \"foo\") { withExec (args: [\"apk\",\"add\",\"--no-cache\",\"gnupg\"]) { stdout  }}}} }\n      "}}
      at makeRequest (file:///private/tmp/repro_mount_dir/node_modules/graphql-request/src/index.ts:441:11)
      at processTicksAndRejections (node:internal/process/task_queues:95:5) {
    response: { data: null, errors: [Array], status: 200, headers: [Headers] },
    request: {
      query: '\n' +
        '        { container  { from (address: "alpine:3.17") { withMountedSecret (path: "/root/.gnupg/pubring.kbx",source: "foo") { withExec (args: ["apk","add","--no-cache","gnupg"]) { stdout  }}}} }\n' +
        '      ',
      variables: undefined
    }
  },
  code: 'D100',
  requestContext: {
    query: '\n' +
      '        { container  { from (address: "alpine:3.17") { withMountedSecret (path: "/root/.gnupg/pubring.kbx",source: "foo") { withExec (args: ["apk","add","--no-cache","gnupg"]) { stdout  }}}} }\n' +
      '      ',
    variables: undefined
  },
  response: {
    data: null,
    errors: [ [Object] ],
    status: 200,
    headers: Headers { [Symbol(map)]: [Object: null prototype] }
  }
}
```

**Comment by grouville** [comment]
I have explored our graphQL implementation. Currently only have a partial implementation of the spec, still some issues on non-printable unicode chars. All the experiments are available [here](https://github.com/grouville/dagger/tree/test-string-graphql/test-string-graphql).

Theorytically, what Alex proposes is possible, still haven't found the implementation yet.

Next step:
```discord
https://pkg.go.dev/github.com/wundergraph/graphql-go-tools@v1.62.3/pkg/astprinter
Alex: curious what this does if you build up a *ast.Document containing a binary string and then print the generated query
we depend on a fork of this package already
```

**Comment by grouville** [comment]
Here is a quick overview of where the issue lies, what the arbitrages are and potential next steps to its resolution.

# Global problem

When trying to mount binary files into Dagger, we encounter encoding issues on invalid UTF-8 sequences.
Repro: https://gist.github.com/grouville/df32b261dcf8bf1a9c7d140f5c18a09b
This gives us the issue:

```shell
Please visit https://dagger.io/help#go for troubleshooting guidance.|
panic: input:1: Syntax Error GraphQL request (1:49) Invalid character escape sequence: \\x.
```

## 1. Isolating the root issue

It turns out that the issue lies in the [dagger/graphql](https://github.com/dagger/graphql) lib that we forked. My first attempt was to follow Alex’s advice to try understanding how / if we could manage this issue at the level of the library.

### 1. The issue arises at the library level

This is the repro showing that the library doesn’t [handle properly the unicodes](https://gist.github.com/grouville/101e500a019ef49640040c99b1214c6c). Here, I create a new scalar, with a corresponding query handle, but we don’t even get at the resolve level. We get the same error as the one we get from the Dagger Go SDK:

```shell
go run ./main_broken.go
2023/05/19 00:34:30 failed to execute graphql operation, errors: [Syntax Error GraphQL request (3:17) Invalid character "\\u0001"

exit status 1
```

### 2. Implementing the GraphQL spec

As @vito suggested, we explored if there was a way to implement the graphQL’s spec on unicode management. It could theoretically work, but I haven’t found a sustainable way to retrieve the exact data I sent that way:

#### 1. gqlgen strings marshalling
`gqlgen` implemented a marshalling / unmarshalling of strings [compliant with the graphQL specs](https://github.com/99designs/gqlgen/blob/v0.17.31/graphql/string.go#L12).

#### 2. Testing gqlgen strings marshalling as a solution
The idea here is to marshal the binary data and see if we have the exact same bytes data at the server level, and as a response. The test string can be one of the following breaking test:
```go
// ASCII control characters
{"\b"},
{"\f"},
{"\n"},
{"\r"},
{"\t"},

// Escaped character
{"\""},
{"\\"},
{"/"},

// Unicode characters
{"\U0001F4A9"}, // Emoji
{"\u2764"},     // Heart

// Mixed input
{"Hello, \nWorld!\r\nYours,\t\"GraphQL\""},
{" # This comment has a \u0A0A multi-byte character."},
{"\"unicode \\u1234\\u5678\\u90AB\\uCDEF\""},
{"こんにちは, 世界!\n \U0001F4A9"},
{"\"Has a фы世界 multi-byte character.\""},
```
Let’s take: `"\u2764"`  as an example.

[The two repros](https://gist.github.com/grouville/3230d414aaed49437035b4e553401477) show that:
##### First run, with `no_escape.go`:
```shell
$ go run ./no_escape.go
2023/05/22 16:03:21 failed to execute graphql operation, errors: [Syntax Error GraphQL request (3:17) Unexpected character "\\u2764".

2: \u0009{
3: \u0009\u0009echo(message: ❤)
                   ^
4: \u0009}
]
exit status 1
```

##### Second run, with `string_escape.go`:
```shell
$ go run ./string_escape.go
ParseLiteral: |"❤"|
ParseLiteral: |"❤"|
Incoming data matches original data
OUTPUT data matches original data
```

> 1. the gqlgen strings escape works: it escapes all unicodes properly:
> 2. We are able to retrieve from the server the original data

#### 3. JSON Unmarshalling is not unicode safe, in Go

However, JSON unmarshalling of some unicode bytes break in Go, and we do rely on that to unmarshall the server response.

Let's take the almost the same code as above (with gqlgen strings escape), and let's try to load my binary gpg key and see if my output data is equivalent to my input data: [code here](https://gist.github.com/grouville/5e712a8353a2784d2a7c94683ccd7888).

```shell
go run ./main.go
Incoming data matches original data
OUTPUT data does not match original data!
```

What is interesting here is that the 'incoming data', in the server is equivalent to the original data. However, when retrieving it on the response, the response data doesn't match the original data.

##### What is happening ?

It turns out that this all comes from the: 
```go
// echoStr is a JSON string, so it needs to be unmarshaled
// this breaks the response unicode data
var unescapedEchoStr string
if err := json.Unmarshal([]byte(echoStr), &unescapedEchoStr); err != nil { 
	log.Fatalf("failed to unmarshal string: %v", err)
}
```

Here are the repros: https://gist.github.com/grouville/8d8ac47ba16e7980e9edca797492eeaf

* The example with the stdlib `stdlib_json.go`, breaks on my gpg key file:
```shell
go run ./stdlib_json.go
The original string and the string obtained from the string areNOT equal
The original bytes and the bytes obtained from the string are not equal
```

* Serializing unicodes with json unmarshal:

The idea was to rely on https://github.com/nspcc-dev/go-ordered-json. This library is being used in several blockchains for two purposes: json unmarshal ordering and serialization of unicode bytes. It [should manage this issue](https://github.com/nspcc-dev/neo-go/pull/2174). 

However, it doesn't work on my example: my local test doesn't pass either. Check `forked_json.go` for the repro:
```shell
go run ./forked_json.go
The original string and the string obtained from the string areNOT equal
The original bytes and the bytes obtained from the string are not equal
```

Now, we are sure that the gqlgen strings escape works, however, there is an issue with the JSON unmarshalling. We do rely on that in our codebase (at least in Go SDK) to process the server's response: https://github.com/dagger/dagger/blob/main/sdk/go/internal/querybuilder/querybuilder.go#L121C10-L148.

### 3. A Base64 Bytes scalar

Now, using base64, JSON unmarshal seems to not only work perfectly, but seems to be easily handling base64 encoded strings. In the `base64.go` example, I don't decode the base64 string:
```shell
go run ./base64.go 
The original string and the string obtained from JSON unmarshalling are equal
```

With the complete scalar example, relying on a base64 encoding also works perfectly: https://gist.github.com/grouville/ae31ef61a192d9afc55334f220d2db3f.

We could even go further to automatically decode the base64 response with a custom json umarshaller:

```go
type Base64Bytes []byte

func (b *Base64Bytes) UnmarshalJSON(data []byte) error {
	var s string
	if err := json.Unmarshal(data, &s); err != nil {
		return err
	}
	decoded, err := base64.StdEncoding.DecodeString(s)
	if err != nil {
		return err
	}
	*b = decoded
	return nil
}
```

## 2. Proposal

Given all this context, and taking into account the fact that I did not miss anything, I see 2 possible implementations, to be defined with you @helderco.

In this issue, and according to the tests, it can be split in 2 sub-issues:
1. Properly encoding strings: https://github.com/matiasinsaurralde/dagger/blob/9c7a60429fe2b7f1167b0a0d1ebed52d3123071d/sdk/go/internal/querybuilder/marshal.go#L68

Basically, this probably only implies to transform the `strconv.Quote` to the gqlgen strings escape (for the Go SDK only), and do the same for other SDKs. It will make us handle unicodes for 99% of use-cases on our SDKs, and is automatically escaped server-side.

2. Add a custom bytes scalar, for all the APIs that could contain random binary data.
I see 2 options, with big implications:
- A custom Bytes scalar, relying on a base64 encoding to `[]byte{}` server-side
- Keep the strings scalar, but base64 encode on the SDK side as `String`. The API would have a `encoding` field. 

What do you think @helderco ? What would be the best way to handle both scalar types on the API ?

cc @matiasinsaurralde @vito @aluzzardi @helderco

**Comment by helderco** [comment]
@grouville Can you test a repro for the other SDKs failing to properly encode the strings? Python uses a third party library for the query builder while node.js has a custom implementation. Maybe they don't have an issue here?

**Comment by grouville** [comment]
> @grouville Can you test a repro for the other SDKs failing to properly encode the strings? Python uses a third party library for the query builder while node.js has a custom implementation. Maybe they don't have an issue here?

You're right, other SDKs do not have this issue for UTF-8 unicodes, but the issue remains for binary files with weird sequences, as explained above.

---
@sipsma, after talking with @helderco, it seems that we used to have a way to secret files inside the containers containers, with a Dagger equivalent to "fsync". If possible, this would enable us to create a new API endpoint to mount a secret file from a path, instead of reading its content as text. Do you believe this is possible ? 

Otherwise, the base64 approach will have to be followed

**Comment by helderco** [comment]
For the base64 approach, I suggest a different method for backwards compatibility, for example: `Query.setSecretFile(name: String!, contents: Base64!)`.

But better if we can avoid the custom scalar as it adds to the SDKs maintenance burden.

**Comment by helderco** [comment]
@grouville Fyi, I just tested the above and it works. Basically creating a new api field that receives a path, and reading the content with Go, without going to buildkit.

**Comment by helderco** [comment]
Re: binary files... I suggest changing the store's data type in `core/store.go`:
```diff
type Store struct {
-  secrets map[string]string
+  secrets map[string][]byte
}
```
So that `GetSecret` returns the original `[]byte` untouched.

Also adapt `AddSecret` to receive `[]byte` and do the string conversion where it's actually needed like `Query.setSecret()` and `Secret.plaintext()`.

**Comment by grouville** [comment]
Just as an update: nothing is blocking the resolution of this issue ATM. Erik and Helder are aligned on the solution, which is the one proposed by Helder right above: https://discord.com/channels/707636530424053791/1113087733074776164

**Comment by gerhard** [comment]
We (+@helderco) would like to get this done for:
- https://github.com/dagger/dagger/discussions/5374

WDYT @grouville?

**Comment by grouville** [comment]
> We (+@helderco) would like to get this done for:
> 
> * [Breaking changes releases - Timeline, Scope & Open Questions #5374](https://github.com/dagger/dagger/discussions/5374)
> 
> WDYT @grouville?

Totally agree. Prioritizing it 🙏
